src/models/losses.py

```diff
@@ -4,6 +4,327 @@
 import torch
 from torch import nn
 
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-locals
+
+
+def generator_loss(
+    logits: dict[str, dict[str, torch.Tensor]],
+    inception_encoder: Any,
+    fake_imgs: torch.Tensor,
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
+    real_imgs: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the loss for the generator.
+
+    Args:
+        logits: Dictionary with fake/real and word-level/uncond/cond logits
+
+        inception_encoder: The inception encoder model.
```

Comment:
Actually yes I thought about it too, but then currently in the main training loop we pass real images to inception encoder and not fake images. Fake images are passed only for loss calculation purposes and hence I have used them here. I believe this might make the training loop a bit clear and clean and would be better in understanding the flow. I feel the current code is abstracting out the loss calculation by doing it inside the loss function. Let me know if you think this is bad

Reply:
again, the idea of loss is essentially like this:
```
def my_awesome_loss(net_out, true_out):
    # some functional op on tensors
    return some_num
```
and in general while training loop you do
```
for epoch in range(epochs):
    net_out = net(x)
    loss = my_awesome_loss(net_out, true_out)
```
notice how you don't pass the network into loss because the moment you make loss dependent on the particular net, it introduces high coupling which is what we don't want

Reply:
right.. that makes sense. So now we have a trade-off here. Like if we pass directly inception features we lose out on abstraction as the logic would then be in training loop, thus making the training loop bit dirty and unclean.

but if we pass directly the network we will be increasing coupling...
so im not sure how should we optimize here, coupling vs abstraction. let me know what do you think and why it would be better and will try to make the changes accordingly.

Reply:
> making the training loop bit dirty and unclean

I would argue that it won't be dirty and unclean. It's just a forward pass through the encoder, which is a natural thing to do within the training loop

Reply:
got it, i have made the changes. you can review and resolve once pushed